src/commands/dockerInstallerBase.ts

```diff
@@ -0,0 +1,67 @@
+import { OSTempFileProvider } from '../debugging/coreclr/tempFileProvider';
+import { ext } from '../extensionVariables';
+import { localize } from '../localize';
+import { streamToFile } from '../utils/httpRequest';
+import LocalOSProvider from '../utils/LocalOSProvider';
+
+export abstract class DockerInstallerBase {
+    protected abstract downloadUrl: string;
+    protected abstract fileExtension: string;
+    protected abstract installationMessage: string;
+    protected abstract getInstallCommand(fileName: string): string;
+
+    public async downloadAndInstallDocker(): Promise<void> {
+        const confirmInstall: string = localize('vscode-docker.commands.DockerInstallerBase.confirm', 'Are you sure you want to install docker on this machine.');
+        const installTitle: string = localize('vscode-docker.commands.DockerInstallerBase.install', 'Install');
+        const downloadingMessage: string = localize('vscode-docker.commands.DockerInstallerBase.downloading', 'Downloading Docker installer...');
+        const downloadCompleteMessage: string = localize('vscode-docker.commands.DockerInstallerBase.downloadCompleteMessage', 'Download completed');
+        let downlaodedFileName: string;
+
+        // no need to check result - cancel will throw a UserCancelledError
+        await ext.ui.showWarningMessage(confirmInstall, { modal: true }, { title: installTitle });
+
+        await vscode.window.withProgress(
+            { location: vscode.ProgressLocation.Notification, title: downloadingMessage },
+            async () => {
+                downlaodedFileName = await this.downloadInstaller();
+                // eslint-disable-next-line @typescript-eslint/no-floating-promises
+                vscode.window.showInformationMessage(downloadCompleteMessage);
+            }
+        );
+
+        const command = this.getInstallCommand(downlaodedFileName);
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        vscode.window.showInformationMessage(this.installationMessage);
+        await this.install(downlaodedFileName, command);
+    }
+
+    private async downloadInstaller(): Promise<string> {
+        const osProvider = new LocalOSProvider();
+        const processProvider = new ChildProcessProvider();
+        const tempFileProvider = new OSTempFileProvider(osProvider, processProvider);
+        let fileName = tempFileProvider.getTempFilename('docker', this.fileExtension);
+        await streamToFile(this.downloadUrl, fileName);
+        return fileName;
+    }
+
+    protected async install(fileName: string, cmd: string): Promise<void> {
+        const fsProvider = new LocalFileSystemProvider();
+        try {
+            const processProvider = new ChildProcessProvider();
+            ext.outputChannel.appendLine(localize('vscode-docker.commands.DockerInstallerBase.downloadCompleteMessage', 'Executing command {0}', cmd));
+            await processProvider.exec(cmd, {});
```

Comment:
Use the terminal for all OSs, don't spawn the process directly.

Reply:
Otherwise, at least use `spawnAsync` directly instead of `ChildProcessProvider`.

Reply:
We decided to user Terminal only where we need user input.
Also what is the reasoning behind not using ChildProcessProvider.

Reply:
Eventually the coreclr things are going to go away or get refactored. I don't want to keep adding to the amount of work that's going to take. `ChildProcessProvider` just calls `spawnAsync` underneath it.

Reply:
I didn't know that we have a plan to get rid of ChildProcessProvider. I will avoid using it then.

Reply:
Yeah. Definitely a lot of it will be reused/refactored, though.